### PR TITLE
Feature/readlogfile

### PR DESCRIPTION
**Pull request by dehann**
Adding `lcm/examples/read_log.c` style LCM log file interaction. New functions are for reading events from LCM logs through a similar API. Writing to LCM log files will have to be done at a later date. The idea is to power through log files much faster during development than having to throttle UDP multicast traffic to your program speed.

Inspired by similar work from @spillai.

### CODE REVIEW

**Comment by dehann** [comment]
Hi @rdeits, Travis tests are passing but codecov not so much. I haven't worked with codecov too much before unfortunately. I followed the `Details` link above, but not quite sure where to go from there. If you could just point me to where to fix the codecov please. Is this an Appveyor thing?

**Comment by dehann** [comment]
Just hold for a moment -- I may have found a bug... Will be right back

**Comment by dehann** [comment]
okay, that should do it. Sorry for the delay.

File: src/ReadLog.jl [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+# export
+#   LCMlog,
+#   close,
+#   readNextEvent,
+#   handle,
+#   subscribe
+
+# also see JuliaLang's RawFD type, but little information found.
+# www.c4learn/c-programming/c-file-structure-and-file-pointer
+mutable struct FILE
```
Comment (rdeits) [inline src/ReadLog.jl]: I'm a bit concerned about trying to exactly mirror the `FILE` implementation from C here. Is this actually necessary? The code never uses any of these fields, so I think we could replace all of this with a `Ptr{Void}` field in `lcm_eventlog_t`.

File: src/ReadLog.jl [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  buffer::Ptr{Cuchar}
+  curp::Ptr{Cuchar}
+  istemp::Cuint # (unsigned is temp)
+end
+
+mutable struct lcm_eventlog_t
+  f::Ptr{FILE}
+  eventcount::Clonglong
+end
+const _LCMlog = Ptr{lcm_eventlog_t}
```
Comment (rdeits) [inline src/ReadLog.jl]: Can we delete this alias? It's only used in a couple of places, and at least one is marked with a comment indicating what the alias stands for. `Ptr{lcm_eventlog_t}` is pretty short and quite clear.

File: src/ReadLog.jl [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  fd::Cuchar
+  flags::Cuint  # (unsigned flags ;)
+  hold::Cuchar
+  buffer::Ptr{Cuchar}
+  curp::Ptr{Cuchar}
+  istemp::Cuint # (unsigned is temp)
+end
+
+mutable struct lcm_eventlog_t
+  f::Ptr{FILE}
```
Comment (rdeits) [inline src/ReadLog.jl]: I think if this is a `Ptr{Void}` then we can lose the whole `FILE` struct

File: src/ReadLog.jl [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  eventcount::Clonglong
```
Comment (rdeits) [inline src/ReadLog.jl]: In `eventlog.h` this is explicitly an `int64_t`, so we should use an `Int64` here as well.

File: src/ReadLog.jl [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+end
+const _LCMlog = Ptr{lcm_eventlog_t}
+
+
+struct lcm_eventlog_event_t
+  eventnum::Clonglong
```
Comment (rdeits) [inline src/ReadLog.jl]: Likewise, can you please replace the `long long`s and `long`s here with the explicit `Int64` and `Int32` from `eventlog.h`?

File: src/ReadLog.jl [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  function LCMlog(filename::S) where {S <: AbstractString}
+    try
+      _log = lcm_eventlog_create(filename)
+      return new(_log, Dict{AbstractString, LCMCore.SubscriptionOptions}())
+    catch e
+      warn(e)
+      backtrace()
+      throw("Cannot open the LCM log file at: $filename")
+    end
+    return new() # should never occur
```
Comment (rdeits) [inline src/ReadLog.jl]: Why is this line necessary?

File: src/ReadLog.jl [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  try
+    lcm_eventlog_destroy(lcmlog._log)
+  catch e
+    warn(e)
+  end
+end
+
+isgood(_event::Ptr{lcm_eventlog_event_t}) = _event != C_NULL
+
+function readNextEvent(lcmlog::LCMlog)::Union{Void, lcm_eventlog_event_t}
```
Comment (rdeits) [inline src/ReadLog.jl]: Can you please make this `underscore_case`?

File: src/ReadLog.jl [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+      throw("Cannot open the LCM log file at: $filename")
+    end
+    return new() # should never occur
+  end
+end
+
+
+
+function close(lcmlog::LCMlog)
+  try
```
Comment (rdeits) [inline src/ReadLog.jl]: What is the purpose of the `try`-catch here? The `ccall` in `lcm_eventlog_destroy` shouldn't throw a Julia error; rather, I'd expect it to segfault if we try to destroy the same log twice, in which case this `try` won't help.

File: src/ReadLog.jl [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+# Doesn't work yet
+# event = unsafe_wrap(lcm_eventlog_event_t, _event, dim??)
+
+
+
+mutable struct LCMlog
+  _log::_LCMlog
+  subscriptions::Dict{AbstractString, LCMCore.SubscriptionOptions}
+  function LCMlog(filename::S) where {S <: AbstractString}
+    try
```
Comment (rdeits) [inline src/ReadLog.jl]: This `try-catch` also seems unnecessary. Opening a nonexistent logfile doesn't actually throw, it just returns a log with its `_log` field set to NULL.

I think that we should remove this `try-catch` block and modify `lcm_eventlog_create` to throw an error if its `ccall` returns NULL.

File: src/ReadLog.jl [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    )
+end
+
+#
+# Doesn't work yet
+# event = unsafe_wrap(lcm_eventlog_event_t, _event, dim??)
+
+
+
+mutable struct LCMlog
```
Comment (rdeits) [inline src/ReadLog.jl]: This should be `LCMLog` for consistency with `LCMType` etc.

**Comment by rdeits** [comment]
Oh and don't worry about the `codecov` status: it just "fails" if the total code coverage goes down too much in a single PR. As long as the tests are running your new code it should be just fine.

File: src/ReadLog.jl (hunk 3 above)
Comment (dehann) [reply src/ReadLog.jl]: Sure, will change back -- I actually started with `Ptr{Void}`. Also not sure if `Base.RawFD` is trying to do the same thing?

File: src/ReadLog.jl (hunk 4 above)
Comment (dehann) [reply src/ReadLog.jl]: Okay, I was under the impression I had to use the C types. I followed the `ccall` [reference for wrapping structs](https://docs.julialang.org/en/stable/manual/calling-c-and-fortran-code/#Mapping-C-Types-to-Julia-1). You suggesting we just go with?
```julia
mutable struct lcm_eventlog_t
  f::Ptr{Void}
  eventcount::Int64
end
```

File: src/ReadLog.jl (hunk 2 above)
Comment (dehann) [reply src/ReadLog.jl]: Sure, I was trying something else and just left it there in the end. Will swap it out.

File: src/ReadLog.jl (hunk 5 above)
Comment (dehann) [reply src/ReadLog.jl]: yes

File: src/ReadLog.jl (hunk 9 above)
Comment (dehann) [reply src/ReadLog.jl]: Can do that.

File: src/ReadLog.jl (hunk 6 above)
Comment (dehann) [reply src/ReadLog.jl]: Old habit... Will remove.

File: src/ReadLog.jl (hunk 8 above)
Comment (dehann) [reply src/ReadLog.jl]: Indeed a segfault. Will use an `if isgood(lcmlog)` instead.

File: src/ReadLog.jl (hunk 7 above)
Comment (dehann) [reply src/ReadLog.jl]: Okay

**Comment by dehann** [comment]
Think I have all of them, and improved the tests a bit too. Thanks for the comments -- let me know if you want me to change anything else.

File: src/ReadLog.jl (hunk 3 above)
Comment (rdeits) [reply src/ReadLog.jl]: Ok, thanks! I think RawFD is wrapping a Unix file descriptor, rather than a C file, but I admit I'm not entirely sure.

File: src/ReadLog.jl (hunk 4 above)
Comment (rdeits) [reply src/ReadLog.jl]: Yeah, exactly. You're right that using the C types is generally the right thing to do, but there's nothing special about them. For example, on my system, `Cint === Int32`. The reason they're recommended is that if you have a C library that uses an `int`, then you want Julia to use the same type, regardless of what `int` means in your particular architecture. But the LCM C library explicitly uses `int64_t` here, so the right way to ensure Julia is using the same type is also to use Int64.

File: src/ReadLog.jl (hunk 2 above)
Comment (rdeits) [reply src/ReadLog.jl]: Thanks!

File: src/ReadLog.jl (hunk 9 above)
Comment (rdeits) [reply src/ReadLog.jl]: Thank you!

File: src/ReadLog.jl (hunk 6 above)
Comment (rdeits) [reply src/ReadLog.jl]: gotcha

File: src/ReadLog.jl (hunk 8 above)
Comment (rdeits) [reply src/ReadLog.jl]: Perfect, thank you!

**Comment by rdeits** [comment]
This is great, thank you again!
